### PR TITLE
[TESTING] Multithread scan kernel

### PR DESCRIPTION
**Pull request by jeanp413**

### CODE REVIEW

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h [hunk 1]
```diff
@@ -268,10 +269,25 @@ class TableScan : public kernel {
...
-			this->add_to_output_cache(std::move(batch));
+		int TABLE_SCAN_KERNEL_NUM_THREADS = 1;
+		std::map<std::string, std::string> config_options = context->getConfigOptions();
+		auto it = config_options.find("TABLE_SCAN_KERNEL_NUM_THREADS");
+		if (it != config_options.end()){
+			TABLE_SCAN_KERNEL_NUM_THREADS = std::stoi(config_options["TABLE_SCAN_KERNEL_NUM_THREADS"]);
 		}
+
+		std::vector<BlazingThread> threads;
+		for (int i = 0; i < TABLE_SCAN_KERNEL_NUM_THREADS; i++) {
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.h:280]: This is alot more overhead than we need. If th elimit is 4 for example we shouldn't have 4 threads running at the same time but rather have up to 4 .next() being called at the same time and limiting another .next() from calling until one of  of them finishes. This is alot more overhead than the pattern having a single thread that launches load_batch as a seperate thread that can finish whenever it needs to

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h [hunk 2]
```diff
@@ -312,30 +328,43 @@ class BindableTableScan : public kernel {
...
+		auto it = config_options.find("TABLE_SCAN_KERNEL_NUM_THREADS");
+		if (it != config_options.end()){
+			TABLE_SCAN_KERNEL_NUM_THREADS = std::stoi(config_options["TABLE_SCAN_KERNEL_NUM_THREADS"]);
+		}
+
+		std::vector<BlazingThread> threads;
+		for (int i = 0; i < TABLE_SCAN_KERNEL_NUM_THREADS; i++) {
+			threads.push_back(BlazingThread([expression = this->expression, this]() {
+				std::unique_ptr<ral::frame::BlazingTable> batch;
+				while(batch = input.next()) {
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.h:343]: Same as other scan, this is much less efficient than having 1 thread for most of the logic and seperate threads for the heavy lifting

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h (hunk 1 above)
Comment (williamBlazing) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.h:280]: i dont understand what overhead you are talking about. It sounds like you are talking about creating a different thread for every file. THat does not sound efficient

**Comment by kharoc** [comment]
1 RAL

========================================================
FAILED TESTS                                                                                                                                                                                                
========================================================                                                                                                                                                    
                        TestGroup InputType Result                                           TestId
0  Bindable Alias                  orc       Fail   TEST_09                                        
1  Date                            orc       Fail   TEST_01                                        
2  Simple Distribution From Local  orc       Fail   TEST_22,TEST_43,TEST_46,TEST_49                
3  TPCH Queries                    orc       Fail   TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
4  Timestamp                       orc       Fail   TEST_02                                        
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1261/1274
FAILED: 13/1274
CRASH: 0/1274
TOTAL: 1274
MAX DELTA: 554.0
*******************************************************************************
Aggregations without group by:   Start Mem: 427.875   End Mem: 549.875   Diff: 122.0
Coalesce:   Start Mem: 549.875   End Mem: 621.875   Diff: 72.0
Column Basis:   Start Mem: 621.875   End Mem: 611.875   Diff: -10.0
Common Table Expressions:   Start Mem: 611.875   End Mem: 815.875   Diff: 204.0
Count without group by:   Start Mem: 815.875   End Mem: 817.875   Diff: 2.0
Date:   Start Mem: 817.875   End Mem: 1165.875   Diff: 348.0
Timestamp:   Start Mem: 1165.875   End Mem: 1167.875   Diff: 2.0
Full outer join:   Start Mem: 1167.875   End Mem: 1167.875   Diff: 0.0
Group by:   Start Mem: 1167.875   End Mem: 1613.875   Diff: 446.0
Group by without aggregations:   Start Mem: 1613.875   End Mem: 1601.875   Diff: -12.0
Inner join:   Start Mem: 1601.875   End Mem: 1975.875   Diff: 374.0
Left outer join:   Start Mem: 1975.875   End Mem: 1965.875   Diff: -10.0
Non-EquiJoin Queries:   Start Mem: 1965.875   End Mem: 2347.875   Diff: 382.0
Nested Queries:   Start Mem: 2347.875   End Mem: 2403.875   Diff: 56.0
Order by:   Start Mem: 2403.875   End Mem: 2501.875   Diff: 98.0
Predicates With Nulls:   Start Mem: 2501.875   End Mem: 2833.875   Diff: 332.0
Simple String:   Start Mem: 2833.875   End Mem: 2933.875   Diff: 100.0
Unary ops:   Start Mem: 3341.875   End Mem: 3439.875   Diff: 98.0
Unify Tables:   Start Mem: 3439.875   End Mem: 3403.875   Diff: -36.0
Union:   Start Mem: 3403.875   End Mem: 3485.875   Diff: 82.0
Limit:   Start Mem: 3485.875   End Mem: 3761.875   Diff: 276.0
Where clause:   Start Mem: 3761.875   End Mem: 3887.875   Diff: 126.0
Bindable Alias:   Start Mem: 3887.875   End Mem: 4269.875   Diff: 382.0
Boolean:   Start Mem: 4269.875   End Mem: 4217.875   Diff: -52.0
Case:   Start Mem: 4217.875   End Mem: 4509.875   Diff: 292.0
Cast:   Start Mem: 4509.875   End Mem: 4955.875   Diff: 446.0
Concat:   Start Mem: 4955.875   End Mem: 5323.875   Diff: 368.0
Literal:   Start Mem: 5323.875   End Mem: 5307.875   Diff: -16.0
Dir:   Start Mem: 5307.875   End Mem: 5861.875   Diff: 554.0
File System Local:   Start Mem: 5861.875   End Mem: 5779.875   Diff: -82.0
File System S3:   Start Mem: 5779.875   End Mem: 5823.875   Diff: 44.0
File System Google Storage :   Start Mem: 5823.875   End Mem: 5905.875   Diff: 82.0
Like:   Start Mem: 5905.875   End Mem: 6249.875   Diff: 344.0
Substring:   Start Mem: 6793.875   End Mem: 6779.875   Diff: -14.0
Wild Card:   Start Mem: 6779.875   End Mem: 7209.875   Diff: 430.0
TPCH Queries:   Start Mem: 7209.875   End Mem: 7675.875   Diff: 466.0
Round:   Start Mem: 7675.875   End Mem: 8091.875   Diff: 416.0

2 Worker's

========================================================
FAILED TESTS                                                                                                                                                                                                
========================================================                                                                                                                                                    
                        TestGroup InputType Result                                           TestId
0  Bindable Alias                  orc       Fail   TEST_09                                        
1  Date                            orc       Fail   TEST_01                                        
2  Simple Distribution From Local  orc       Fail   TEST_22,TEST_43,TEST_46,TEST_49                
3  TPCH Queries                    orc       Fail   TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
4  Timestamp                       orc       Fail   TEST_02                                        
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1500/1513
FAILED: 13/1513
CRASH: 0/1513
TOTAL: 1513
MAX DELTA: 473.0
*******************************************************************************
Aggregations without group by:   Start Mem: 844.875   End Mem: 1317.875   Diff: 473.0
Coalesce:   Start Mem: 1317.875   End Mem: 1359.875   Diff: 42.0
Common Table Expressions:   Start Mem: 1359.875   End Mem: 1383.875   Diff: 24.0
Count without group by:   Start Mem: 1383.875   End Mem: 1429.875   Diff: 46.0
Date:   Start Mem: 1429.875   End Mem: 1445.875   Diff: 16.0
Timestamp:   Start Mem: 1445.875   End Mem: 1455.875   Diff: 10.0
Full outer join:   Start Mem: 1455.875   End Mem: 1453.875   Diff: -2.0
Group by:   Start Mem: 1453.875   End Mem: 1519.875   Diff: 66.0
Group by without aggregations:   Start Mem: 1519.875   End Mem: 1571.875   Diff: 52.0
Inner join:   Start Mem: 1571.875   End Mem: 1543.875   Diff: -28.0
Left outer join:   Start Mem: 1543.875   End Mem: 1637.875   Diff: 94.0
Non-EquiJoin Queries:   Start Mem: 1637.875   End Mem: 1687.875   Diff: 50.0
Nested Queries:   Start Mem: 1687.875   End Mem: 1831.875   Diff: 144.0
Order by:   Start Mem: 1831.875   End Mem: 1823.875   Diff: -8.0
Predicates With Nulls:   Start Mem: 1823.875   End Mem: 1959.875   Diff: 136.0
Simple String:   Start Mem: 1959.875   End Mem: 1963.875   Diff: 4.0
Unary ops:   Start Mem: 2131.875   End Mem: 2199.875   Diff: 68.0
Unify Tables:   Start Mem: 2199.875   End Mem: 2173.875   Diff: -26.0
Union:   Start Mem: 2173.875   End Mem: 2127.875   Diff: -46.0
Limit:   Start Mem: 2127.875   End Mem: 1995.875   Diff: -132.0
Where clause:   Start Mem: 1995.875   End Mem: 1993.875   Diff: -2.0
Bindable Alias:   Start Mem: 1993.875   End Mem: 2123.875   Diff: 130.0
Boolean:   Start Mem: 2123.875   End Mem: 2161.875   Diff: 38.0
Case:   Start Mem: 2161.875   End Mem: 2195.875   Diff: 34.0
Cast:   Start Mem: 2195.875   End Mem: 2353.875   Diff: 158.0
Concat:   Start Mem: 2353.875   End Mem: 2487.875   Diff: 134.0
Literal:   Start Mem: 2487.875   End Mem: 2415.875   Diff: -72.0
Dir:   Start Mem: 2415.875   End Mem: 2763.875   Diff: 348.0
File System Local:   Start Mem: 2763.875   End Mem: 2699.875   Diff: -64.0
File System S3:   Start Mem: 2699.875   End Mem: 2759.875   Diff: 60.0
File System Google Storage :   Start Mem: 2759.875   End Mem: 2803.875   Diff: 44.0
Like:   Start Mem: 2803.875   End Mem: 2965.875   Diff: 162.0
Substring:   Start Mem: 3285.875   End Mem: 3339.875   Diff: 54.0
Wild Card:   Start Mem: 3339.875   End Mem: 3435.875   Diff: 96.0
TPCH Queries:   Start Mem: 3435.875   End Mem: 3669.875   Diff: 234.0
Round:   Start Mem: 3669.875   End Mem: 3677.875   Diff: 8.0

**Pull request merged**